src/linux_udev.rs

```diff
@@ -0,0 +1,762 @@
+//! Functions for talking to udev directly instead of going through hidapi
+
+extern crate udev;
+
+use std::{
+    cell::{Cell, Ref, RefCell},
+    convert::TryInto,
+    ffi::{CStr, CString, OsStr, OsString},
+    fs::{File, OpenOptions},
+    io::Read,
+    os::{
+        fd::{AsRawFd, OwnedFd},
+        unix::{ffi::OsStringExt, fs::OpenOptionsExt},
+    },
+    path::{Path, PathBuf},
+    sync::Mutex,
+};
+
+use nix::{
+    errno::Errno,
+    ioctl_read, ioctl_readwrite_buf,
+    poll::{poll, PollFd, PollFlags},
+    unistd::{read, write},
+};
+
+use super::{BusType, DeviceInfo, HidDeviceBackend, HidError, HidResult, WcharString};
+
+/// Global error to simulate what C hidapi does
+static GLOBAL_ERROR: Mutex<RefCell<Option<String>>> = Mutex::new(RefCell::new(None));
```

Comment:
You really do not need to emulate the C API, it would probably be a better idea to fix the API and make a breaking change. Why make the effort of implementing this in rust, and then keep bad C API practices ;-)

I imagine it would be pretty simple to remove check_error() API's when the udev backend is enabled.  For any actual legitimate use of this API, better to return the errors directly in all functions via Result.

Reply:
I agree, I was wondering if it might be too forward to propose a PR to deprecate it, since it doesn't seem to return anything that you wouldn't have in a better way.

We could deprecate it in mainline and make it a no-op or just not even have it with non-C-hidapi backends if and when they come in. Mostly I added it because one of the examples uses it, so it would actually compile and run. If it goes away selectively then probably that would have to change.